src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -91,46 +70,83 @@ public BoundLambda(SyntaxNode syntax, BoundBlock body, ImmutableArray<Diagnostic
             );
         }
 
-        public TypeSymbolWithAnnotations InferredReturnType(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        public TypeSymbolWithAnnotations GetInferredReturnType(ref HashSet<DiagnosticInfo> useSiteDiagnostics, NullableWalker.VariableState nullableState = null)
         {
-#if DEBUG
-            Debug.Assert(_hasInferredReturnType);
-#endif
-
-            if (!_inferredReturnTypeUseSiteDiagnostics.IsNullOrEmpty())
+            if (!InferredReturnType.UseSiteDiagnostics.IsEmpty)
             {
                 if (useSiteDiagnostics == null)
                 {
                     useSiteDiagnostics = new HashSet<DiagnosticInfo>();
                 }
-
-                foreach (var info in _inferredReturnTypeUseSiteDiagnostics)
+                foreach (var info in InferredReturnType.UseSiteDiagnostics)
                 {
                     useSiteDiagnostics.Add(info);
                 }
             }
-
-            return _inferredReturnType;
+            if (nullableState == null)
+            {
+                return InferredReturnType.Type;
+            }
+            else
+            {
+                var returnTypes = ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)>.GetInstance();
+                // Diagnostics from NullableWalker.Analyze can be dropped here since Analyze
```

Comment:
>Diagnostics from NullableWalker.Analyze can be dropped here [](start = 19, length = 59)

Seems wasteful, since the diagnostics are the same ones we would compute in the caller (I think). Moreover, it sounds like it could get exponential with nested lambdas.

Reply:
Added comment.

---
In reply to: [192914856](https://github.com/dotnet/roslyn/pull/27283#discussion_r192914856) [](ancestors = 192914856)